ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/validate/DeduplicationState.scala

```diff
@@ -51,15 +53,80 @@ case class DeduplicationState private (
 }
 
 object DeduplicationState {
-  private[sandbox] type DeduplicationQueue = VectorMap[ChangeId, Time.Timestamp]
+  private[sandbox] type DeduplicationQueue = UnsafeDeduplicationStateQueueMap
 
   private[validate] def empty(
       deduplicationDuration: Duration,
       bridgeMetrics: BridgeMetrics,
   ): DeduplicationState =
     DeduplicationState(
-      deduplicationQueue = VectorMap.empty,
+      deduplicationQueue = UnsafeDeduplicationStateQueueMap.empty,
       maxDeduplicationDuration = deduplicationDuration,
       bridgeMetrics = bridgeMetrics,
     )
+
+  /** This data structure is tailored for keeping an ordered (by insertion) sequence of deduplication entries
+    * of the form (changeId, recordTime) with optimal complexity of each exposed operation.
+    *
+    * @param vector An ordered (by insertion) vector of deduplication entries
+    * @param mappings Mapping of changeId to recordTime
+    */
+  private[validate] case class UnsafeDeduplicationStateQueueMap(
```

Comment:
I am not sure what you are referring to with the Unsafe here: this implementation seems to be completely thread-safe, since immutable all the way :thinking: 
Can it be if Unsafe not needed here?

Reply:
You're right, not needed.